menhir-recover/synthesis.ml

```diff
@@ -290,18 +287,62 @@ struct
-    and print_actions ppf = Utils.pp_list print_action ppf
+    and print_actions ppf = Utils.pp_list print_action ppf in
+    let cost_to_string cost =
+      if Cost.is_infinite cost then "inf" else sprintf "%2d" (Cost.to_int cost) in
+    let print_item_with_solution ppf = function
+      | (prod, pos), (cost, actions) ->
+         fprintf ppf "Item(%3d, %1d) at cost %s:\n %a => %a\n"
+             (Production.to_int prod) pos (cost_to_string cost)
+             Print.item (prod, pos)
+             print_actions actions
     in
-    List.iter (fun (item, states) ->
-        fprintf ppf "# Item (%d,%d)\n"
-          (Production.to_int (fst item)) (snd item);
-        Print.item ppf item;
-        List.iter (fun ((cost, actions), states) ->
-            fprintf ppf "at cost %d from states %a:\n%a\n\n"
-              (cost : Cost.t :> int)
-              (Utils.pp_list (fun ppf st ->
-                   fprintf ppf "#%d" (Lr1.to_int st))) states
-              print_actions actions
-          ) (group_assoc states)
-      ) (group_assoc solutions)
+    let sort_by_cost =
+      List.stable_sort (fun (_, (cost1, _)) (_, (cost2, _)) -> compare cost1 cost2) in
+    fprintf ppf "\n\n\n(* * COSTS OF SOLUTIONS: *)\n\n";
+    List.iter (fun (state, items) ->
+        fprintf ppf "State: #%d\n" (Lr1.to_int state);
+        fprintf ppf "\n";
+        List.iter (fun (item, solution) ->
+                print_item_with_solution ppf (item, solution))
+             (sort_by_cost items);
+        fprintf ppf "\n\n"
+      ) solutions;
+    fprintf ppf "\n(* * COST OF PRODUCTIONS (not equal one): *)\n\n";
+    Lr1.iter (fun lr1 ->
+      List.iter (fun (prod, pos) ->
+        let prod_len = Array.length (Production.rhs prod) in
+        let cost = cost_of_prod prod in
+        if pos = prod_len && Cost.compare cost (Cost.of_int 1) <> 0 then
+            fprintf ppf "Production %a at cost %s\n"
+                Print.production prod
+                (cost_to_string cost)
+      ) (Lr0.items (Lr1.lr0 lr1))
+    );
+    fprintf ppf "\n(* * COST OF TERMINAL (not equal one) : *)\n\n";
+    Terminal.iter (fun t ->
+      let cost = cost_of_symbol (T t) in
+      if (Cost.compare cost (Cost.of_int 1)) <> 0 then
+        fprintf ppf "%a at cost %s\n"
+            Print.terminal t
+            (cost_to_string cost)
+    );
+    fprintf ppf "\n(* * COST OF NONTERMINAL (not equal inf) : *)\n\n";
+    Nonterminal.iter (fun n ->
+        let cost = cost_of_symbol (N n) in
+        if not (Cost.is_infinite cost) then
+          fprintf ppf "%a at cost %s\n"
+              Print.nonterminal n
+              (cost_to_string @@ cost_of_symbol (N n))
+    );
+    fprintf ppf "\n(* * PENALTY OF ITEMS (not equal zero): *)\n\n";
+    Lr0.iter (fun lr0 ->
+      List.iter (fun ((prod, pos) as item) ->
+         let cost = penalty_of_item item in
+         if (Cost.compare cost Cost.zero) <> 0 then
+             fprintf ppf "Item (%3d, %d) at cost %s\n"
```

Comment:
Use `Print.item`. `prod` id is also useful for better navigation

Reply:
fixed